### PR TITLE
testing: use pytest idioms for unit tests

### LINKED ISSUE #498: Refactor unit tests to use py.test idioms
For future code habitability.

### PR DESCRIPTION
**Pull request by tseaver**
Closes #498.

### CODE REVIEW

**Comment by tseaver** [comment]
Reviewers:  this is a big PR, and mostly involves repetition of the same few transforms:

- Use normal `assert` rather than `self.assert*`.
- Replace `self.assertRaises` with `pytest.raises`.
- Convert testcase helper methods to module-scope functions.
- Drop use of the `unittest.TestCase` base class, converting testcase methods to module-scope functions.
- Minor cleanups of module-scope imports.

**Comment by tseaver** [comment]
4ff4506 seems to have broken `Watch.on_snapshot` -- I will check it out tomorrow.

All the earlier "test refactoring" commits should be reviewable as it.

**Comment by tseaver** [comment]
I've rebased the branch to remove the last two commits (4ff4506 and 10eb32f), leaving this PR as purely unit-test refactoring.

**Review by tswast — CHANGES_REQUESTED** [review]
A few minor issues, but otherwise LGTM

File: tests/unit/v1/test__helpers.py [hunk 1]
```diff
@@ -13,2492 +13,2540 @@
...
+    assert not inst.empty_document
+    assert not inst.has_transforms
+
+
+def test_documentextractor_get_update_pb_w_exists_precondition():
+    from google.cloud.firestore_v1.types import write
+
+    document_data = {}
+    inst = _make_document_extractor(document_data)
+    document_path = "projects/project-id/databases/(default)/" "documents/document-id"
```
[the remark this answers is not in the document]
Comment (tseaver) [reply tests/unit/v1/test__helpers.py]: 866bf1d for this and ff.

File: tests/unit/v1/test_transforms.py [hunk 2]
```diff
@@ -12,102 +12,105 @@
...
+    from google.cloud.firestore_v1.transforms import _NumericValue
+
+    return _NumericValue(*args, **kwargs)
+
+
+def test__numericvalue_ctor_w_invalid_types():
+    invalid_values = (None, u"phred", b"DEADBEEF", [], {}, object())
+    for invalid_value in invalid_values:
+        with pytest.raises(ValueError):
+            _make_numeric_value(invalid_value)
```
[the remark this answers is not in the document]
Comment (tseaver) [reply tests/unit/v1/test_transforms.py]: e723d5d

The first argument needs to be a comma-delimited string.

File: tests/unit/v1/test_transforms.py [hunk 3]
```diff
@@ -12,102 +12,105 @@
...
+    for value in values:
+        inst = _make_numeric_value(value)
+        assert inst.value == value
+
+
+def test__numericvalue_ctor_w_float():
+    values = (-10.0, -1.0, 0.0, 1.0, 10.0)
+    for value in values:
+        inst = _make_numeric_value(value)
+        assert inst.value == value
```
[the remark this answers is not in the document]
Comment (tseaver) [reply tests/unit/v1/test_transforms.py]: e723d5d

File: tests/unit/v1/test_base_client.py [hunk 4]
```diff
@@ -13,424 +13,466 @@
...
-class TestBaseClient(unittest.TestCase):
 
-    PROJECT = "my-prahjekt"
+def _make_base_client(*args, **kwargs):
+    from google.cloud.firestore_v1.base_client import BaseClient
 
-    @staticmethod
-    def _get_target_class():
-        from google.cloud.firestore_v1.client import Client
+    return BaseClient(*args, **kwargs)
```
[the remark this answers is not in the document]
Comment (tseaver) [reply tests/unit/v1/test_base_client.py:27]: Nope -- the test was actually not working as intended (it was copy-pasta from the three-way split).

File: tests/unit/v1/test_base_client.py [hunk 5]
```diff
@@ -13,424 +13,466 @@
...
-class Test__parse_batch_get(unittest.TestCase):
-    @staticmethod
-    def _call_fut(get_doc_response, reference_map, client=mock.sentinel.client):
-        from google.cloud.firestore_v1.base_client import _parse_batch_get
+def test__reference_info():
+    from google.cloud.firestore_v1.base_client import _reference_info
 
-        return _parse_batch_get(get_doc_response, reference_map, client)
+    expected_doc_paths = ["/a/b", "/a/b/c/d", "/a/b", "/f/g"]
+    documents = [mock.Mock(_document_path=path) for path in expected_doc_paths]
```
[the remark this answers is not in the document]
Comment (tseaver) [reply tests/unit/v1/test_base_client.py:348]: I'd prefer not:  unit tests should rely as little as possible on other modules from the same package.

File: tests/unit/v1/test_client.py [hunk 6]
```diff
@@ -14,473 +14,524 @@
...
-        self.assertEqual(collection1._path, parts)
-        self.assertIs(collection1._client, client)
-        self.assertIsInstance(collection1, CollectionReference)
 
-        # Make sure using segments gives the same result.
-        collection2 = client.collection(*parts)
-        self.assertEqual(collection2._path, parts)
-        self.assertIs(collection2._client, client)
-        self.assertIsInstance(collection2, CollectionReference)
+def test_client__firestore_api_property():
```
[the remark this answers is not in the document]
Comment (tseaver) [reply tests/unit/v1/test_client.py:75]: I'm pretty sure I added it to provide explicit coverage for the property, which previously was covered only incidentally (part of that `Client` vs. `BaseClient` bit above).

File: tests/unit/v1/test_query.py [hunk 7]
```diff
@@ -13,762 +13,776 @@
...
+        request={
+            "parent": parent_path,
+            "structured_query": query._to_protobuf(),
+            "transaction": None,
+        },
+        metadata=client._rpc_metadata,
+    )
+
+
+_not_passed = object()
```
[the remark this answers is not in the document]
Comment (tseaver) [reply tests/unit/v1/test_query.py:536]: Cleaning out module-scope imports from the test module:  the `gapic_v1` module is now imported inside the helper function.

File: tests/unit/v1/test_base_document.py [hunk 8]
```diff
@@ -12,412 +12,421 @@
...
+
+    # We use != explicitly since assertEqual would use ==.
+    assert not (document1 != document4)
+    assert document1 is not document4
+
+
+def test_basedocumentreference__ne__other_type():
+    document = _make_base_document_reference("X", "YY", client=mock.sentinel.client)
+    other = object()
+    document != other
```
[the remark this answers is not in the document]
Comment (tseaver) [reply tests/unit/v1/test_base_document.py]: f888567

File: tests/unit/v1/test_bundle.py [hunk 9]
```diff
@@ -458,50 +506,56 @@ def test_deserialize_from_seconds_nanos(self):
         # instead of seconds/nanos.
         # re_serialized = bundle.build()
         # # Finally, confirm the round trip.
-        # self.assertEqual(
-        #     re_serialized,
-        #     _helpers.deserialize_bundle(re_serialized, client=client).build(),
-        # )
+        # assert re_serialized == _helpers.deserialize_bundle(re_serialized, client=client).build()
```
[the remark this answers is not in the document]
Comment (tseaver) [reply tests/unit/v1/test_bundle.py:511]: 1ba0f41617ad95e5ef5d07acf0d63aa12397a8a2

File: tests/unit/v1/test_client.py [hunk 10]
```diff
@@ -14,473 +14,524 @@
...
-        transaction = client.transaction(max_attempts=3, read_only=True)
-        self.assertIsInstance(transaction, Transaction)
-        self.assertEqual(transaction._write_pbs, [])
-        self.assertEqual(transaction._max_attempts, 3)
-        self.assertTrue(transaction._read_only)
-        self.assertIsNone(transaction._id)
 
+def test_client_recursive_delete_raises():
+    client = _make_default_client()
+    pytest.raises(TypeError, client.recursive_delete, object())
```
[the remark this answers is not in the document]
Comment (tseaver) [reply tests/unit/v1/test_client.py]: Nope. 5c84656

File: tests/unit/v1/test_query.py [hunk 11]
```diff
@@ -13,762 +13,776 @@
...
+        "snooze", direction=firestore.Query.DESCENDING
+    ).limit_to_last(2)
+    returned = query.get()
+
+    assert isinstance(returned, list)
+    assert query._orders[0].direction == _enum_from_direction(firestore.Query.ASCENDING)
+    assert len(returned) == 2
+
+    snapshot = returned[0]
+    assert snapshot.reference._path == ("dee", "sleep")
```
[the remark this answers is not in the document]
Comment (tseaver) [reply tests/unit/v1/test_query.py:139]: 25b32f9

File: tests/unit/v1/test_transaction.py [hunk 12]
```diff
@@ -12,1010 +12,994 @@
...
-        sleep.assert_called_once_with(uniform.return_value)
+        ),
+    ]
+    assert firestore_api.begin_transaction.mock_calls == expected_calls
+    firestore_api.rollback.assert_not_called()
+    commit_call = mock.call(
+        request={"database": db_str, "writes": [], "transaction": txn_id},
+        metadata=transaction._client._rpc_metadata,
+    )
+    assert firestore_api.commit.mock_calls, [commit_call == commit_call]
```
[the remark this answers is not in the document]
Comment (tseaver) [reply tests/unit/v1/test_transaction.py]: acf9f8d

**Review by tswast — APPROVED** [review]

**Pull request merged**